mixedversion: run skip-version upgrades when supported

DarrylWong opened the pull request:
This changes the `mixedversion` framework to randomly perform "skip-version" upgrades when supported. Specifically, v23.1 onwards supports skip-version upgrades, where a single rolling upgrade can be performed to upgrade Cockroach two major releases at a time.

This behaviour is still experimental, and it is necessary to set an environment variable to enable this feature.

Tests that perform logic specific to a certain upgrade path can disable this functionality by using the `DisableSkipVersionUpgrades` option or gating certain steps behind a `helper.IsSkipVersionUpgrade` check.

Fixes: #121957
Epic: none
Release note: None

Commit:
mixedversion: run skip-version upgrades when supported

This changes the `mixedversion` framework to randomly perform
"skip-version" upgrades when supported. minSupportedSkipVersionUpgrade
indicates the minimum version after which skip version upgrades
are supported. The diff between the current version and the
minSupportedSkipVersionUpgrade denotes the number of possible
version skips.

Specifically, v24.1 onwards supports skip-version upgrades,
where a single rolling upgrade can be performed to upgrade
Cockroach two major releases at a time.

Tests that perform logic specific to a certain upgrade path can
disable this functionality by using the `DisableSkipVersionUpgrades`
option or gating certain steps behind a `helper.IsSkipVersionUpgrade`
check.

Fixes: #121957
Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125869)
<!-- Reviewable:end -->

DarrylWong commented:
[Test runs of mvt tests here](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/15752347?buildTab=tests#%2Fcdc%2Fmixed-versions%2Frun_2%2Fartifacts.zip)

One failure that has been fixed on master, and two related to removal of [non-mux rangefeed code](https://github.com/cockroachdb/cockroach/commit/294e45d7ee6060c83a8d44e38d75162911205d4d#diff-e0f73599f690bbfe7af43a737b0719186ff22308003907f994f3f438adfde45d) that is still being investigated but shouldn't affect this PR.

renatolabs commented:
🎉 

A couple of changes related to the backup-restore test:

* it also runs the `schemachange` workload, so I presume we'll need the same logic for skip version upgrades.
* it restarts / wipes the cluster to test cluster restores. So we'll need to pass the `ALLOW_VERSION_SKIPPING` env var there as well in order for the restore to succeed.

DarrylWong commented:
>it also runs the schemachange workload, so I presume we'll need the same logic for skip version upgrades.

Correct me if I'm wrong, but although the `schemaChangeWorkloadCmd` helper lives in `mixed_version_backup.go`, it's not actually called by the mixed version test, only by `registerBackupRestoreRoundTrip` (the non mvt backup-restore tests).

> So we'll need to pass the ALLOW_VERSION_SKIPPING env var

Ah, good catch. Added, but currently running some tests to make sure it worked.

renatolabs commented:
> the non mvt backup-restore tests

Huh, for the longest time I've been thinking that we run `schemachange` on that test but yeah, we don't. It seems there was an intention to do it (#115062), but it never happened. Never mind then!

I also see a test failure in one of the data driven tests. Suggestion: I think it would be nice to disable version skipping by default on the datadriven tests, and add a separate test that enables it unconditionally (so that we're able to test it in isolation).

renatolabs commented on pkg/cmd/roachtest/tests/mixed_version_backup.go:
```diff
@@ -2441,7 +2445,7 @@ func (mvb *mixedVersionBackup) verifyAllBackups(
 
 			if _, ok := collection.btype.(*clusterBackup); ok {
-				err := u.resetCluster(ctx, l, version, h.ExpectDeaths)
+				err := u.resetCluster(ctx, l, version, h.ExpectDeaths, mixedversion.ClusterSettingsForUpgrade(h.Context().FromVersion, h.Context().ToVersion))
```
Hrmm, I wonder if exposing this function is worth it. We could either unconditionally pass the env var or compute whether we need the env var locally using `h.IsSkipVersionUpgrade`.

Reason I say this is because it reduces (even if a little) the API we expose. And backup-restore is pretty unique and the only test that does this kind of restart.

If you do want to keep this logic, we should fix it to just pass the env var when actually needed. Currently, it _seems_ that the code is doing that, but it's also passing the env var when the `version` argument to the `verify` function is `ToVersion`, and that's unnecessary.

DarrylWong replied:
SGTM! Reverted exposing the function and went with `h.IsSkipVersionUpgrade` approach.

renatolabs commented:
> It seems there was an intention to do it (https://github.com/cockroachdb/cockroach/pull/115062), but it never happened.

cc @msbutler, in case there's any interest from DR and this fell through the cracks.

msbutler commented:
> cc @msbutler, in case there's any interest from DR and this fell through the cracks.

Yeah, it's been on my backlog. The schemachange workload flakes enough in the non mixed version test 😄

DarrylWong commented:
>I also see a test failure in one of the data driven tests.

Ah, it's because I added a testing commit to unconditionally enable version skipping.

>disable version skipping by default on the datadriven tests

Done!

>add a separate test that enables it unconditionally

One already exists! Added by you in `skip_version_upgrades` 😄

renatolabs commented:
This LGTM, but I'll let others review.

srosenberg commented on pkg/testutils/release/releases.go:
```diff
@@ -56,6 +56,32 @@ func parseReleases() (map[string]Series, error) {
+// the two versions passed. Returns an error when there is no
+// predecessor information for a release in the chain (which should
+// only happen if one of the versions passed is very old).
+func MajorReleasesBetween(v1, v2 *version.Version) (int, error) {
```
Nit: `v1` -> `from`, `v2` -> `to`.

srosenberg replied:
Nm, since the order isn't required, maybe update the comment to say `...between _any_ two versions passed`.

DarrylWong replied:
Done.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/clusterupgrade/clusterupgrade.go:
```diff
@@ -87,6 +87,11 @@ func (v *Version) AtLeast(other *Version) bool {
 }
 
+// Series returns the release series this version is a part of.
+func (v *Version) Series() string {
```
Duplicates `versionSeries` in `release`. Maybe we should just make the latter public and reuse?

DarrylWong replied:
Done.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/mixedversion/helper.go:
```diff
@@ -297,6 +298,17 @@ func (h *Helper) loggerFor(name string) (*logger.Logger, error) {
+// N.B. Calling this in a startup step may lead to unexpected results,
+// as there is no upgrade going on.
+func (h *Helper) IsSkipVersionUpgrade() bool {
+	numReleases, _ := release.MajorReleasesBetween(&h.testContext.FromVersion().Version, &h.testContext.ToVersion().Version)
```
We can ignore the error because both `FromVersion` and `ToVersion` are validated elsewhere?

DarrylWong replied:
Stealing Renato's comment (and also adding it to the code above):

```
	// N.B.: it is safe to ignore the error here as we should always be
	// able to determine the number of releases between two binaries
	// used when upgrading: we keep release data starting from 21.2.
```

From my understanding that reasoning would apply here as well.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go:
```diff
@@ -428,6 +430,12 @@ func DisableMutators(names ...string) CustomOption {
+// minSupportedSkipVersionUpgrade is the minimum version after which
+// "skip version" upgrades are supported (i.e., upgrading two major
+// releases in a single upgrade). Note that support is currently
+// experimental and requires an environment variable.
```
Nit: might as well mention the name of that environment variable.

Should we add a `TODO` here for the note? Presumably, it won't remain experimental indefinitely.

DarrylWong replied:
If we are no longer testing 23.2 skipping, do we need the environment variable anymore? Or is it still "experimental" in 24.1 -> 24.3 (do we still need the env var)?

renatolabs replied:
We can get rid of the env var logic.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go:
```diff
@@ -678,33 +704,102 @@ func (t *Test) runCommandFunc(nodes option.NodeListOption, cmd string) stepFunc
+		}
+
+		chosenPredecessor := predecessors[t.prng.Intn(len(predecessors))]
+		upgradePath = append(upgradePath, chosenPredecessor)
```
Nit: might be worth asserting that it maintains a strict descending order; it's not immediately obvious,

```
assert(len(upgradePath) == 1 || ugradePath[len(upgradePath]-1] < chosenPredecessor`
```

renatolabs replied:
IMO, this invariant is part of the contract of the `releases` package (i.e., when you call `releases.LatestPredecessor`, you get an immediate predecessor). If we wanted to add an assertion, I'd rather assert ordering when loading the releases file (`parseReleases`).

renatolabs commented:
Internal discussion [link](https://cockroachlabs.slack.com/archives/C05JB0QURND/p1719959975155199); TLDR: we're no longer testing 23.2 -> 24.2 upgrades.

We'll need to set `minSupportedSkipVersionUpgrade` to 24.1.0. It'll make the changes in this PR effectively invisible for the time being. However, once master becomes 24.3, tests should start performing skip-version upgrades.

DarrylWong commented on pkg/testutils/release/releases.go:
```diff
@@ -56,6 +56,32 @@ func parseReleases() (map[string]Series, error) {
+		count++
+		seriesData, ok := releaseData[currentSeries]
+		if !ok {
+			return -1, fmt.Errorf("no release data for release series %s", currentSeries)
```
`currentSeries` is constructed using `v.Major` and `v.Minor` which are both ints. So worst case it would show `v0.0`, but I added a `:` anyway to make it clear there should be something there.

srosenberg commented on pkg/cmd/roachtest/tests/versionupgrade.go:
```diff
@@ -155,6 +155,13 @@ func runVersionUpgrade(ctx context.Context, t test.Test, c cluster.Cluster) {
+			// The schema change workload is only compatible with the branch it was built
+			// from and the major version before that.
+			if h.IsSkipVersionUpgrade() {
+				l.Printf("skipping: upgrade is a version skip")
```
Nit: "version skip upgrades are unsupported"

DarrylWong replied:
Switched to "schema change workload is unsupported with skip version upgrades"

srosenberg commented on pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go:
```diff
@@ -74,17 +75,16 @@ func TestTestPlanner(t *testing.T) {
 
 	datadriven.Walk(t, datapathutils.TestDataPath(t, "planner"), func(t *testing.T, path string) {
-		resetMutators()
+		eraseMutators()
```
Nit: I actually liked `resetMutators` :) `eraseXXX` usually denotes something more involved than clearing a pointer. Maybe `clearMutators` instead?

DarrylWong replied:
I'm okay with `resetMutators` :). I think the name is still "resetting" the mutators but `clearMutators` works with me too if others like that more.

srosenberg commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mutators_test.go:
```diff
@@ -37,18 +37,20 @@ func TestPreserveDowngradeOptionRandomizerMutator(t *testing.T) {
+		require.Equal(t, mutationInsertBefore, mutations[j+1].op)
+		require.IsType(t, allowUpgradeStep{}, mutations[j+1].impl)
+
+		j += 2 // check next pair
```
Nit: we should update the comment for `mut.Generate` to say that the length of the resulting sequence is always even.

DarrylWong replied:
That is only true for the `preserveDowngradeOptionRandomizerMutator` implementation I think? Unless you meant just for that specific implementation, in which case I've added it.

srosenberg reviewed (approved):
Nice work! LGTM other than the comment,

> Specifically, v23.1 onwards supports skip-version upgrades, where a single rolling upgrade can be performed to upgrade Cockroach two major releases at a time.

That should now be `24.1`. I also think we should explicitly mention `minSupportedSkipVersionUpgrade` in the commit summary, since that's essentially a feature flag; i.e., diff between current version and minSupportedSkipVersionUpgrade denotes the number of potential version skips.

DarrylWong commented:
>I also think we should explicitly mention minSupportedSkipVersionUpgrade in the commit summary, since that's essentially a feature flag

Done, also cleaned up the commit message of the outdated env var comment.

DarrylWong commented:
TFTRs!

bors r=srosenberg, renatolabs

renatolabs commented on pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:
```diff
@@ -421,7 +421,9 @@ func (p *testPlanner) changeVersionSteps(
 		steps = append(steps, p.newSingleStep(
-			restartWithNewBinaryStep{version: to, node: node, rt: p.rt, settings: p.clusterSettings()},
+			restartWithNewBinaryStep{
+				version: to, node: node, rt: p.rt,
```
@DarrylWong seems like we're no longer passing the cluster settings when restarting binaries here. I think this is a bug?

DarrylWong replied:
Ah yes thank you, that's definitely not correct. Will need to test it out but might be why [change-replicas/mixed-version](https://github.com/cockroachdb/cockroach/issues/127106) is failing. Still chasing another bug related to the datadriven test thats blocking the backports (I think I just need to bump the version). I'll put out a fix for them.

renatolabs replied:
Yes, we're not passing the `install.EnvOption{"COCKROACH_SCAN_MAX_IDLE_TIME=10ms"}` anymore which is likely why the test started to fail.
